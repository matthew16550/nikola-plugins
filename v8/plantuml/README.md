This plugin converts [PlantUML](https://plantuml.com/) files to SVG, PNG or TXT (Ascii Art) files.

There is also a [plantuml_markdown](https://plugins.getnikola.com/#plantuml_markdown) plugin which supports
PlantUML inside Markdown files.

The default configuration will output all `*.puml` files under the `pages` dir as SVG files.

Developed against PlantUML version 1.2020.23.  Probably works with some earlier versions.

# Unicode

The plugin expects PlantUML files to be encoded with UTF-8.

# Known Issues

- It's slow!  Every PlantUML rendering uses a new Java process, on my laptop it takes 4-8 seconds per file.
  I have some ideas to speed this up, and they may be available in future plugin versions.

- Changes to files included via `!include ...` will NOT trigger a rebuild.
  Instead, you could include them via `PLANTUML_ARGS` and those will trigger a rebuild.

- `nikola auto` does not watch dirs in `PLANTUML_FILES` or files included via `PLANTUML_ARGS` / `!include`.
  As a workaround you could put PlantUML files under any dir listed in `POSTS` or `PAGES` because those dirs
  are watched.
  (Use `.iuml` suffix for include files to prevent them matching the `*.puml` wildcard in `PLANTUML_FILES`)

- The line number in PlantUML error messages is one higher than the actual error location.
  Suspect a bug in PlantUML.

- The file name in PlantUML error messages is always `string` rather than the actual problem file.
  PlantUML does this when input is piped via stdin, which as a compromise for simplicity we always do.
