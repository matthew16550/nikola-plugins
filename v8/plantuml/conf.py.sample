#
# PLANTUML_EXEC (list of strings) - The command to run PlantUML
#
# '%site_path%' anywhere in PLANTUML_EXEC will be replaced with the full path to the site dir.
# PlantUML is run in the site dir so often this is not needed.
#
# Examples
# --------
# Run from a JAR file:
#   [ 'java', '-Djava.awt.headless=true', '-jar', 'plantuml.jar' ]
#
# Run in Docker:
#   [ 'docker', 'run', '--interactive', '--rm', '--volume', '%site_path%:/work', 'YOUR_IMAGE',
#     'java', '-Djava.awt.headless=true', '-jar', 'plantuml.jar' ]
#
PLANTUML_EXEC = ['plantuml']

#
# PLANTUML_ARGS (list of strings) - CLI arguments that are sent to PlantUML, see https://plantuml.com/command-line
#
# Examples
# --------
# Use a common style file for all diagrams:
#   [ '-Imy_plantuml_style.iuml' ]
#
# Specify the style in conf.py
#   [ '-chide footbox', '-SShadowing=false' ]
#
PLANTUML_ARGS = []

#
# PLANTUML_FILES contains (wildcard, destination, output_format, args) tuples.
#
# <wildcard> is used to generate a list of source files (in the same way as POSTS and PAGES).
#
# <output_format> is one of 'png', 'svg', 'txt'.
#
# Rendered files will be placed at:
#   output / <destination> / <filename>.<output_format>
#
# As with POSTS and PAGES you can create any directory structure you want and it will be reflected in the output.
#
# <args> is a list of cli arguments that are appended to PLANTUML_ARGS
#
PLANTUML_FILES = (
    ('pages/*.puml', '', 'svg', []),
)

#
# PLANTUML_DEBUG (boolean) - Control verbosity of this plugin
#
PLANTUML_DEBUG = False

#
# PLANTUML_RENDER_ERRORS (boolean) - If True then PlantUML error messages will be rendered in the target files.
#
# You might find this helpful when running `nikola auto`.
#
PLANTUML_RENDER_ERRORS = False
