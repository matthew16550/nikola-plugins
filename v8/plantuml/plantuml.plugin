[Core]
Name = plantuml
Module = plantuml
Tests = test_plantuml

[Nikola]
MinVersion = 8.0.4
PluginCategory = Task

[Documentation]
Author = Matthew Leather
Version = 0.1
Website = https://plugins.getnikola.com/#plantuml
Description = Renders PlantUML files
