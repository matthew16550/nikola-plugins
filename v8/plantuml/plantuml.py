import os
import subprocess
from logging import DEBUG
from math import ceil
from pathlib import Path
from typing import List, Optional
from xml.sax.saxutils import escape

from nikola import Nikola, utils
from nikola.plugin_categories import Task

DEFAULT_ARGS = []

DEFAULT_DEBUG = False

DEFAULT_EXEC = ['plantuml']

DEFAULT_FILES = (
    ('pages/*.puml', '', 'svg', []),
)

DEFAULT_RENDER_ERRORS = True

FORMAT_ARG = {
    'png': '-tpng',
    'svg': '-tsvg',
    'txt': '-tutxt',
}


class PlantUmlTask(Task):
    """Renders PlantUML files"""

    name = 'plantuml'

    def __init__(self):
        super().__init__()
        self.plantuml_manager: Optional[PlantUmlManager] = None

    def set_site(self, site):
        super().set_site(site)
        self.plantuml_manager = PlantUmlManager(site)

    def gen_tasks(self):
        yield self.group_task()

        common_args = self.plantuml_manager.common_args
        filters = self.site.config['FILTERS']
        output_folder = self.site.config['OUTPUT_FOLDER']
        plantuml_files = self.site.config.get('PLANTUML_FILES', DEFAULT_FILES)
        output_path = Path(output_folder)

        # Logic derived from nikola.plugins.misc.scan_posts.ScanPosts.scan()
        for pattern, destination, output_format, args in plantuml_files:
            if output_format not in ['png', 'svg', 'txt']:
                raise Exception("Unsupported PlantUML output format '{}'".format(output_format))
            pattern = Path(pattern)
            root = pattern.parent
            combined_args = common_args + args
            includes = [a[2:] for a in combined_args if a.startswith('-I')]
            kw = {
                'combined_args': combined_args,
                'filters': filters,
                'output_folder': output_folder,
            }

            for src in root.rglob(pattern.name):
                dst = output_path / destination / src.relative_to(root).parent / src.with_suffix('.' + output_format).name
                dst_str = str(dst)
                task = {
                    'basename': self.name,
                    'name': dst_str,
                    'file_dep': includes + [src],
                    'targets': [dst_str],
                    'actions': [(self.render_file, [src, dst, output_format, combined_args])],
                    'uptodate': [utils.config_changed(kw, 'plantuml:' + dst_str)],
                    'clean': True,
                }
                yield utils.apply_filters(task, filters)

    def render_file(self, src: Path, dst: Path, output_format, args: List[str]) -> bool:
        output = self.plantuml_manager.render(src.read_bytes(), output_format, args)
        dst.parent.mkdir(parents=True, exist_ok=True)
        dst.write_bytes(output)
        return True


class PlantUmlManager:
    """
    PlantUmlManager is used by "plantuml" & "plantuml_markdown" plugins
    """

    def __init__(self, site: Nikola):
        self.common_args = site.config.get('PLANTUML_ARGS', DEFAULT_ARGS)
        self._exec = site.config.get('PLANTUML_EXEC', DEFAULT_EXEC)
        self._logger = utils.get_logger('plantuml_manager')
        if site.config.get('PLANTUML_DEBUG', DEFAULT_DEBUG):
            self._logger.level = DEBUG
        self._render_errors = site.config.get('PLANTUML_RENDER_ERRORS', DEFAULT_RENDER_ERRORS)
        self._site = site

    # noinspection PyBroadException
    def render(self, source: bytes, output_format: str, args: List[str]) -> bytes:
        _exec = [e.replace('%site_path%', os.getcwd()) for e in self._exec]
        command = _exec + args + ['-pipe', '-stdrpt', FORMAT_ARG[output_format]]
        self._logger.debug('render() exec: %s\n%s', command, source)
        result = subprocess.run(command, capture_output=True, input=source)

        if result.returncode == 0:
            return result.stdout

        try:
            details = str(result.stderr, encoding='utf8').rstrip()
        except Exception:
            details = str(result.stderr)

        error = "PlantUML error (return code {}): {}".format(result.returncode, details)

        if not self._render_errors:
            raise Exception(error)

        self._logger.exception('Ignoring: %s', error)

        if output_format == 'png':
            return error_to_png(error)
        elif output_format == 'svg':
            return error_to_svg(error)
        else:
            return error.encode('utf8')


def error_to_png(error: str) -> bytes:
    import io
    from PIL import Image, ImageDraw
    size = ImageDraw.Draw(Image.new('RGB', (1, 1))).multiline_textsize(error)
    image = Image.new('RGB', (size[0] + 20, size[1] + 20), color=(0xdc, 0x35, 0x45))
    draw = ImageDraw.Draw(image)
    draw.multiline_text((10, 10), error, fill=(255, 255, 255))
    buf = io.BytesIO()
    image.save(buf, 'PNG')
    return buf.getvalue()


def error_to_svg(error: str) -> bytes:
    # This whole function is kludgy because I don't know SVG very well.  Improvements welcome :-)
    lines = error.split('\n')
    text = '\n'.join('<tspan x="1em" dy="1.2em">{}</tspan>'.format(escape(line)) for line in lines)
    height = ceil(1.2 * (len(lines))) + 2
    width = max(len(line) for line in lines) + 2
    result = '<?xml version="1.0" encoding="UTF-8" standalone="no"?>\n' + \
             '<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="{}em" height="{}em">\n'.format(width, height) + \
             '<rect x="0" y="0" width="{}em" height="{}em" fill="#dc3545"/>\n'.format(width, height) + \
             '<text x="0" y="1em" font-family="monospace" fill="#ffffff">\n' + \
             text + '\n' + \
             '</text>\n' + \
             '</svg>'
    return result.encode('utf8')
