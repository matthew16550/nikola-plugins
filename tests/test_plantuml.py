import os
from pathlib import Path
from shutil import copyfile
from typing import Dict

import pytest
from approvaltests import ApprovalException, FileApprover, get_default_reporter
from approvaltests.pytest.namer import PyTestNamer
from pytest import fixture

from tests.base import execute_plugin_tasks
from v8.plantuml.plantuml import PlantUmlTask

BAD_PLANTUML = """
A -> B
B -> C
foo bar
""".lstrip()

GOOD_PLANTUML = """
A -> B : test \u2713
A -> B : defined=$defined
""".lstrip()

CHECK_MARK = '\u2713'  # for unicode testing


def test_render_png_file(verify_plantuml_file):
    verify_plantuml_file('skinparam dpi 300\n' + GOOD_PLANTUML, 'foo', 'png')  # TODO 300 is a temporary kludge


def test_render_png_file_error(verify_plantuml_file):
    verify_plantuml_file(BAD_PLANTUML, 'foo', 'png')


def test_render_svg_file(verify_plantuml_file):
    verify_plantuml_file(GOOD_PLANTUML, '', 'svg')


def test_render_svg_file_error(verify_plantuml_file):
    verify_plantuml_file(BAD_PLANTUML, '', 'svg')


def test_render_txt_file(verify_plantuml_file):
    verify_plantuml_file(GOOD_PLANTUML, 'foo/bar', 'txt')


def test_render_txt_file_error(verify_plantuml_file):
    verify_plantuml_file(BAD_PLANTUML, 'foo/bar', 'txt')


@fixture
def verify_plantuml_file(monkeypatch, tmp_path, verify_file):
    def f(text, destination, output_format):
        monkeypatch.chdir(tmp_path)

        (tmp_path / 'pages').mkdir()
        (tmp_path / 'pages' / 'test.puml').write_text(text, encoding='utf8')
        (tmp_path / 'pages' / 'includes').mkdir()
        (tmp_path / 'pages' / 'includes' / 'include1.iuml').write_text('A -> B : this is include1 ' + CHECK_MARK, encoding='utf8')
        (tmp_path / 'pages' / 'includes' / 'include2.iuml').write_text('A -> B : this is include2 ' + CHECK_MARK, encoding='utf8')

        plugin = create_plugin({
            'PLANTUML_ARGS': [
                '-c!$defined="should-be-overridden-by-specific-define"',
                '-chide footbox',
                '-nometadata',
                '-Ipages/includes/include1.iuml',
                '-SDefaultFontName=DejaVu Sans',
                '-SShadowing=false',
            ],
            'PLANTUML_FILES': (
                ('pages/*.puml', destination, output_format, [
                    '-c!$defined="FOO \u2713"',
                    '-Ipages/includes/include2.iuml',
                ]),
            ),
            'PLANTUML_RENDER_ERRORS': True,
        })

        # When
        execute_plugin_tasks(plugin)

        # Then
        verify_file((tmp_path / 'output' / destination / 'test').with_suffix('.' + output_format))

    return f


@pytest.mark.parametrize('config, expected_targets', [
    # Test the default config
    (
            {},
            [
                'output/test1.svg',
                'output/foo/test2.svg',
            ]
    ),
    # Test multiple rules & subdirs
    (
            {
                'PLANTUML_FILES': (
                        ('pages/*.puml', '', 'txt', []),
                        ('other/diagrams/*.puml', 'foo/bar', 'png', []),
                ),
            },
            [
                'output/test1.txt',
                'output/foo/test2.txt',
                'output/foo/bar/test3.png',
                'output/foo/bar/baz/test4.png',
            ]
    )
])
def test_gen_tasks_targets(config, expected_targets, monkeypatch, tmp_path):
    monkeypatch.chdir(tmp_path)

    for f in [
        'pages/test1.puml',
        'pages/foo/test2.puml',
        'other/diagrams/test3.puml',
        'other/diagrams/baz/test4.puml',
    ]:
        (tmp_path / f).parent.mkdir(parents=True, exist_ok=True)
        (tmp_path / f).touch()

    plugin = create_plugin(config)

    # When
    tasks = plugin_tasks(plugin)

    # Then
    assert sorted((task['targets'][0] for task in tasks), reverse=True) == expected_targets


def test_gen_tasks_deps(monkeypatch, tmp_path):
    monkeypatch.chdir(tmp_path)

    plugin = create_plugin({
        'PLANTUML_ARGS': [
            '-Iincludes/include1.iuml',
            '-Iincludes/include2.iuml'
        ],
        'PLANTUML_FILES': (
            ('*.puml', '', 'txt', ['-Iincludes/include3.iuml']),
        )
    })

    (tmp_path / 'foo.puml').touch()

    # When
    tasks = plugin_tasks(plugin)

    # Then
    assert tasks[0]['file_dep'] == [
        'includes/include1.iuml',
        'includes/include2.iuml',
        'includes/include3.iuml',
        Path('foo.puml'),
    ]


def create_plugin(config: Dict):
    plugin = PlantUmlTask()
    plugin.set_site(FakeSite(config))
    return plugin


class FakeSite:
    debug = True

    def __init__(self, config: Dict):
        self.config = {
            'FILTERS': {},
            'OUTPUT_FOLDER': 'output',
            'PLANTUML_DEBUG': True,
        }
        self.config.update(config)
        if 'PLANTUML_EXEC' in os.environ:
            self.config['PLANTUML_EXEC'] = os.environ['PLANTUML_EXEC'].split()


def plugin_tasks(plugin):
    tasks = list(plugin.gen_tasks())
    assert tasks.pop(0) == plugin.group_task()
    return tasks


@fixture
def verify_file(request):
    def f(path: Path):
        namer = PyTestNamer(request, path.suffix)
        approved = namer.get_approved_filename()
        received = namer.get_received_filename()
        copyfile(path, received)
        if not FileApprover().verify_files(approved, received, get_default_reporter()):
            raise ApprovalException("Approval Mismatch")

    return f
