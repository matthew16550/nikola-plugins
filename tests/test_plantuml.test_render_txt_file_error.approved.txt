PlantUML error (return code 200): string:4:error:Syntax Error?